field/TinyLang/Field/Generator.hs

```diff
@@ -228,81 +227,126 @@ arbitraryUnOpRing = elements [Neg]
 arbitraryBinOpRing :: MonadGen m => m (BinOp f (AField f) (AField f) (AField f))
 arbitraryBinOpRing = elements [Add, Sub, Mul]
 
-groundArbitraryFreqs
-    :: (Field f, Arbitrary f, KnownUni f a, MonadGen m)
+groundArbitraryFreqs :: (Field f, Arbitrary f, KnownUni f a, MonadGen m)
     => Vars f -> [(Int, m (Expr f a))]
 groundArbitraryFreqs vars =
     [ (1, EConst <$> arbitraryM)
     , (2, EVar   <$> chooseUniVar vars)
     ]
 
--- | Generate an expression of a particular type from a collection of variables
+newtype SGenT f (m :: Type -> Type) a =
+    SGen { unSGenT :: GenT (SupplyT (StateT (Vars f) m)) a }
+    deriving newtype ( Monad
+                     , Functor
+                     , Applicative
+                     , MonadSupply
+                     , MonadState (Vars f)
+                     , MonadGen
+                     )
+
+-- TODO:  Is there a way to automate this?
+instance MonadState s m => MonadState s (GenT m) where
```

Comment:
You could define a newtype, provide an instance for it and then use `DerivingVia`. But who cares.